zebra-state/src/service/queued_blocks.rs

```diff
@@ -196,4 +196,80 @@ impl QueuedBlocks {
     pub fn utxo(&self, outpoint: &transparent::OutPoint) -> Option<transparent::Utxo> {
         self.known_utxos.get(outpoint).cloned()
     }
+
+    /// Clears known_utxos, by_parent, and by_height, then drains blocks.
+    /// Returns all key-value pairs of blocks as an iterator
+    pub fn drain(&mut self) -> Drain<'_, block::Hash, QueuedNonFinalized> {
+        self.known_utxos.clear();
+        self.known_utxos.shrink_to_fit();
+        self.by_parent.clear();
+        self.by_parent.shrink_to_fit();
+        self.by_height.clear();
+
+        self.blocks.drain()
+    }
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct SentHashes {
+    /// A Vec of growable ring bufs to be used for pruning outdated hashes that are
+    /// inserted in the order of their height in a given batch.
+    sent_bufs: Vec<VecDeque<(block::Hash, block::Height)>>,
+
+    /// The growable ring buf to be used for pruning outdated hashes that are mostly
+    /// inserted in the order of their height for this batch of sent hashes
+    curr_buf: VecDeque<(block::Hash, block::Height)>,
```

Comment:
Nitpick: we try not to repeat details that are already in the code itself.

```suggestion
    /// A list of previously sent block batches, each batch is in increasing height order.
    /// We use this list to efficiently prune outdated hashes that are at or below the finalized tip.
    sent_bufs: Vec<VecDeque<(block::Hash, block::Height)>>,

    /// The list of blocks sent in the current batch, in increasing height order.
    curr_buf: VecDeque<(block::Hash, block::Height)>,
```